ledger/participant-integration-api/src/main/scala/platform/store/dao/events/ContractWitnessesTable.scala

```diff
@@ -3,64 +3,40 @@
 
 package com.daml.platform.store.dao.events
 
+import java.sql.Connection
+
 import anorm.{BatchSql, NamedParameter}
 import com.daml.platform.store.Conversions._
 import com.daml.platform.store.DbType
 
-private[events] sealed abstract class ContractWitnessesTable {
-
-  protected val insert: String
+private[events] abstract class ContractWitnessesTable {
 
   protected val TableName = "participant_contract_witnesses"
   protected val IdColumn = "contract_id"
   protected val WitnessColumn = "contract_witness"
 
-  private def toNamedParameter(idAndParty: (ContractId, Party)) =
-    idAndParty match {
-      case (id, party) =>
-        List[NamedParameter](
-          IdColumn -> id,
-          WitnessColumn -> party,
-        )
-    }
-
-  private def prepareBatchInsert(witnesses: WitnessRelation[ContractId]): Option[BatchSql] =
-    batch(insert, Relation.flatten(witnesses).map(toNamedParameter).toSeq)
-
   protected val delete = s"delete from $TableName where $IdColumn = {$IdColumn}"
 
-  private def prepareBatchDelete(ids: Seq[ContractId]): Option[BatchSql] =
+  protected def prepareBatchDelete(ids: Seq[ContractId]): Option[BatchSql] =
     batch(delete, ids.map(id => List[NamedParameter](IdColumn -> id)))
 
   def toExecutables(
       info: TransactionIndexing.ContractWitnessesInfo
-  ): ContractWitnessesTable.Executables = {
-    ContractWitnessesTable.Executables(
-      deleteWitnesses = prepareBatchDelete(info.netArchives.toList),
-      insertWitnesses = prepareBatchInsert(info.netVisibility),
-    )
-  }
-
+  ): ContractWitnessesTable.Executables
 }
 
 private[events] object ContractWitnessesTable {
 
-  final case class Executables(deleteWitnesses: Option[BatchSql], insertWitnesses: Option[BatchSql])
+  trait Executable {
+    def execute()(implicit connection: Connection): Unit
+  }
+
+  final case class Executables(deleteWitnesses: Option[BatchSql], insertWitnesses: Executable)
```

Comment:
I guess you needed this nested `Executable` for `insertWitnesses` for being able to abstract over the different types of Anorm statements (`BatchSql` and `SimpleSql` - one used for H2 and one for Postgresql). Is there a penalty for using `BatchSql` instead of `SimpleSql` for PostgreSQL inserts with unnesting?

Reply:
Yes, that's true. BatchSql and SimpleSql have different class hierarchy, so I needed some wrapper here.